### PR TITLE
Add Suggestions for User Input Errors

### PR DESCRIPTION
**Pull request by plypaul**
### Description

This PR adds suggestions to error messages for metrics / group-by-items specified by the user.

### CODE REVIEW

File: metricflow/query/query_resolver.py [hunk 1]
```diff
@@ -311,22 +339,32 @@ def resolve_query(self, resolver_input_for_query: ResolverInputForQuery) -> Metr
         if resolve_order_by_result.input_to_issue_set_mapping.has_issues:
             mappings_to_merge.append(resolve_order_by_result.input_to_issue_set_mapping)
 
-        # Resolve where.
-        filter_spec_lookup = self._resolve_where(
-            resolution_dag=resolution_dag,
-        )
+        # Resolve all where filters in the DAG.
+        filter_spec_lookup = self._build_filter_spec_lookup(resolution_dag)
+        for filer_spec_resolution in filter_spec_lookup.spec_resolutions:
```
Comment (courtneyholcomb) [inline metricflow/query/query_resolver.py:344]: typo: `filer_spec_resolution` -> `filter_spec_resolution`

File: metricflow/query/resolver_inputs/query_resolver_inputs.py [hunk 2]
```diff
@@ -117,31 +119,65 @@ def ui_description(self) -> str:
...
+                    [where_filter.where_sql_template for where_filter in self.where_filter_intersection.where_filters]
+                )
             )
-            + ")"
+            + "\n)"
         )
 
 
+@dataclass(frozen=True)
+class VirtualResolverInputForWhereFilterIntersection(MetricFlowQueryResolverInput):
```
Comment (courtneyholcomb) [inline metricflow/query/resolver_inputs/query_resolver_inputs.py:142]: Why "Virtual"? Confused what that's supposed to mean in this context 🤔

**Review by courtneyholcomb — APPROVED** [review]
A couple nits but otherwise looks good!

**Review by tlento — COMMENTED** [review]
I'm only like halfway through. Hopefully @courtneyholcomb can pick this up and finish the review, but if she doesn't have time I can get to it later this afternoon.

**Comment by tlento** [comment]
Thanks for the lack of state notifications there, GitHub.....

File: metricflow/query/query_resolver.py (hunk 1 above)
Comment (plypaul) [reply metricflow/query/query_resolver.py]: Updated.

File: metricflow/query/resolver_inputs/query_resolver_inputs.py (hunk 2 above)
Comment (plypaul) [reply metricflow/query/resolver_inputs/query_resolver_inputs.py]: NVM, thought about it more and removed that prefix.

File: metricflow/query/query_parser.py [hunk 3]
```diff
@@ -262,26 +262,25 @@ def _error_message(
...
-                issue_counter += 1
-                issue_set_lines.extend(
-                    [
-                        f"Error #{issue_counter}:\n",
-                        error_issue.ui_description(resolver_input),
-                    ]
-                )
+                lines.append(f"\nError #{issue_counter}:")
+                issue_set_lines: List[str] = [
+                    "Query Input:\n",
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/query/query_parser.py]: Swapped the order. With the query input being on the outside, the nesting seemed a little deep,, and given it's unlikely an input would have multiple issues, the reduction in nesting seemed worthwhile.

File: metricflow/specs/patterns/whitelist_pattern.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+from typing import Sequence
+
+from typing_extensions import override
+
+from metricflow.specs.patterns.spec_pattern import SpecPattern
+from metricflow.specs.specs import InstanceSpec
+
+
+class WhitelistSpecPatten(SpecPattern):
+    """A spec pattern that matches based on a configured whitelist of specs.
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/specs/patterns/whitelist_pattern.py]: Updated.

File: metricflow/query/similarity.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+import logging
+from dataclasses import dataclass
+from typing import Sequence
+
+"""From a previous commit:
+
+Importing fuzzywuzzy causes warning due to non-installed levenshstein package. Suppress this by catching
+the warning during the import in query_parser because this package is not required.
+"""
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/query/similarity.py]: Updated.

**Pull request merged**